### PR TITLE
`ClosedAuditID` coherency check and clean-up scripts

### PR DESCRIPTION
**Pull request by pmamut**
S-70222

### CODE REVIEW

File: fixes/fix-asset-closure-incoherency.sql [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+update dbo.BaseAsset set ClosedAuditID=null where ClosedAuditID is not null and AssetState<128
+select @rowcount=@@ROWCOUNT, @error=@@ERROR
+if @error<>0 goto ERR
+print @rowcount + ' open BaseAsset records cleared'
+
+update dbo.BaseAsset
+set ClosedAuditID=bac.AuditBegin
+from dbo.BaseAsset ba
+cross apply (select top 1 bac.* from BaseAsset bac where bac.ID=ba.ID and bac.AssetState=128 and bac.AuditBegin<=ba.AuditBegin and not exists (select * from BaseAsset bao where bao.ID=bac.ID and (bao.AssetState<128 or (bao.AssetState>=192 and bao.AssetState<255)) and bao.AuditBegin>bac.AuditBegin and bao.AuditBegin<ba.AuditBegin) order by AuditBegin) bac
```
Comment (spazmodius) [inline fixes/fix-asset-closure-incoherency.sql:31]: `order by AuditBegin desc`?

File: fixes/fix-asset-closure-incoherency.sql [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+where ba.AssetState>=128 and (ba.AssetState<192 or ba.AssetState=255) and (ba.ClosedAuditID is null or ba.ClosedAuditID<>bac.AuditBegin)
```
Comment (spazmodius) [inline fixes/fix-asset-closure-incoherency.sql:32]: It seems to me that _every_ closed `ba` will be it's own `bac`. Am I figuring right?

**Comment by spazmodius** [comment]
OK, I think I understand.

You are looking, for each `ba`, for the _earliest_ closed without an intervening open.  In that case, this clause is not doing anything: `and bac.AuditBegin<=ba.AuditBegin`; 

I was thinking of the _most recent_ closed not immediately preceded by a closed.

**Comment by pmamut** [comment]
> You are looking, for each `ba`, for the _earliest_ closed without an intervening open.

Yes.

> In that case, this clause is not doing anything: `and bac.AuditBegin<=ba.AuditBegin`;

It prevents subsequent closures from getting picked up for an initial closure in cases where an asset is closed, re-opened, and then closed again.

File: fixes/fix-asset-closure-incoherency.sql (hunk 2 above)
Comment (pmamut) [reply fixes/fix-asset-closure-incoherency.sql:32]: Only if its `ClosedAuditID` is out of whack - `(ba.ClosedAuditID is null or ba.ClosedAuditID<>bac.AuditBegin)`

File: fixes/fix-asset-closure-incoherency.sql [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+update dbo.BaseAsset_Now
+set ClosedAuditID=ba.ClosedAuditID
+from dbo.BaseAsset ba
+where ba.ID=BaseAsset_Now.ID and ba.AuditEnd is null
+ and ((ba.ClosedAuditID is null and BaseAsset_Now.ClosedAuditID is not null) or (ba.ClosedAuditID is not null and BaseAsset_Now.ClosedAuditID is null) or ba.ClosedAuditID<>BaseAsset_Now.ClosedAuditID)
+select @rowcount=@@ROWCOUNT, @error=@@ERROR
+if @error<>0 goto ERR
+print @rowcount + ' BaseAsset_Now records synced'
+
+alter table dbo.BaseAsset_Now enable trigger all
```
Comment (spazmodius) [inline fixes/fix-asset-closure-incoherency.sql:46]: triggers should be re-enabled before the possible `goto ERR`

**Comment by pmamut** [comment]
Both cases above do what they do because we had never expected closed assets to get deleted - `AssetCannotDeleteClosed` (https://github.com/versionone/Core/blame/ba286d09deb316bda07a93b51c0eb16ea57300da/VersionOne.Domain/Meta.xml#L3377-L3380) had existed since the beginning of time.

However, despite this existing restriction on explicit deletion of closed assets, previously-closed workitems routinely get deleted via `WorkitemDeleteDescendants` cascade rule (https://github.com/versionone/Core/blob/HEAD/VersionOne.Domain/Rules/WorkitemDeleteDescendants.cs#L20-L40) when their open parent workitem is deleted, and child workitems deleted by parent's deletion propagation get returned to their prior state upon parent's un-deletion via `WorkitemUndeleteDescendants` cascade rule (https://github.com/versionone/Core/blob/HEAD/VersionOne.Domain/Rules/WorkitemDeleteDescendants.cs#L57-L83). Both of these actions happen without ever touching `ClosedAuditID` of affected workitems or triggering `SetClosedAudit` rule. This results in closed->deleted->closed transitions while still preserving initial `ClosedAuditID` value of these workitems, as expected.

I've logged [D-16967](https://www7.v1host.com/V1Production/assetdetail.v1?Number=D-16967) to reconcile two cases above with expectation of preserving `ClosedAuditID` from initial closure through a closed->deleted->closed transition.

File: fixes/fix-asset-closure-incoherency.sql (hunk 3 above)
Comment (pmamut) [reply fixes/fix-asset-closure-incoherency.sql:46]: This explicit re-enable is the happy path. Initial `alter table dbo.BaseAsset_Now disable trigger all` automatically gets rolled back by any previous failure that does `goto ERR`.

**Pull request merged**